packages/vite/src/node/plugins/html.ts

```diff
@@ -955,19 +956,41 @@ export function htmlEnvHook(config: ResolvedConfig): IndexHtmlTransformHook {
   const pattern = /%(\S+?)%/g
   const envPrefix = resolveEnvPrefix({ envPrefix: config.envPrefix })
   const env: Record<string, any> = { ...config.env }
+  const ignoredImportMetaEnvKeys = new Set()
+
   // account for user env defines
   for (const key in config.define) {
     if (key.startsWith(`import.meta.env.`)) {
       const val = config.define[key]
-      env[key.slice(16)] = typeof val === 'string' ? val : JSON.stringify(val)
+      if (typeof val === 'string') {
+        try {
+          env[key.slice(16)] = extractStringLiteral(val)
+        } catch {
+          ignoredImportMetaEnvKeys.add(key.slice(16))
+        }
```

Comment:
IIUC `extractStringLiteral` would prevent usages like:

```js
define: {
  'import.meta.env.NULL_STR': 'null'
}
```

(It's a bit of a stretch though 😅)

I wonder if doing:
```suggestion
      if (typeof val === 'string') {
        try {
          env[key.slice(16)] = JSON.parse(val)
        } catch {
          env[key.slice(16)] = val
        }
```

Would be good enough to prevent the breaking change. I've also made a [stackblitz](https://stackblitz.com/edit/vitejs-vite-6qsg8y?file=vite.config.js,main.js,index.html&terminal=dev) to test things out. Looking back it seems to be a hard feature to balance, and I could be persuaded that the original issue (https://github.com/vitejs/vite/issues/13424) could be working as intended too.

Reply:
```js
define: {
  'import.meta.env.DOUBLE_QUOTE': "'foo'",
  'import.meta.env.SINGLE_QUOTE': "'foo'",
  'import.meta.env.BACK_QUOTE': "`foo`"
}
```

If we use `JSON.parse`, `import.meta.env.SINGLE_QUOTE` will be replaced with `'foo'`. But I think people will expect that to be replaced with `foo` like `import.meta.env.DOUBLE_QUOTE` works.

|  | `extractStringLiteral` + fallback | `JSON.parse` + fallback |
|-|---|-----|
|DOUBLE_QUOTE| `foo` | `foo` |
|SINGLE_QUOTE| `foo` | `'foo'` |
|BACK_QUOTE| `foo` | `` `foo` `` |

Reply:
I think that's fine for now to reduce complexity, they can workaround it by using `"\"foo\""` (which is usually what `JSON.stringify` generates) 🤔 I'm mostly concern of the additional code here that's needed to maintain for a (seemingly) simple feature.